File: sdks/go/pkg/verinode/internal/http.go
package internal

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/Great-2025/verinode-go/pkg/verinode"
)

// HTTPClient implements the HTTPClient interface
type HTTPClient struct {
	baseURL    string
	httpClient *http.Client
	authToken  string
	maxRetries int
	retryDelay time.Duration
	backoff    float64
}

// NewHTTPClient creates a new HTTP client
func NewHTTPClient(config verinode.ConfigProvider) verinode.HTTPClient {
	return &HTTPClient{
		baseURL: config.GetAPIEndpoint(),
		httpClient: &http.Client{
			Timeout: time.Duration(config.GetTimeout()) * time.Millisecond,
		},
		maxRetries: config.GetMaxRetries(),
		retryDelay: time.Duration(config.GetRetryDelay()) * time.Millisecond,
		backoff:    config.GetBackoffMultiplier(),
	}
}

// Get performs a GET request
func (c *HTTPClient) Get(ctx context.Context, endpoint string, response interface{}, params ...string) error {
	url := c.baseURL + endpoint
	if len(params) > 0 {
		url += "?" + params[0]
	}
	
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	
	return c.doRequestWithRetry(req, response)
}

// Post performs a POST request
func (c *HTTPClient) Post(ctx context.Context, endpoint string, request interface{}, response interface{}) error {
	url := c.baseURL + endpoint
	
	var body io.Reader
	if request != nil {
		jsonData, err := json.Marshal(request)
		if err != nil {
			return fmt.Errorf("failed to marshal request: %w", err)
		}
		body = bytes.NewBuffer(jsonData)
	}
	
	req, err := http.NewRequestWithContext(ctx, "POST", url, body)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	
	if request != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	
	return c.doRequestWithRetry(req, response)
}

// Patch performs a PATCH request
func (c *HTTPClient) Patch(ctx context.Context, endpoint string, request interface{}, response interface{}) error {
	url := c.baseURL + endpoint
	
	var body io.Reader
	if request != nil {
		jsonData, err := json.Marshal(request)
		if err != nil {
			return fmt.Errorf("failed to marshal request: %w", err)
		}
		body = bytes.NewBuffer(jsonData)
	}
	
	req, err := http.NewRequestWithContext(ctx, "PATCH", url, body)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	
	if request != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	
	return c.doRequestWithRetry(req, response)
}

// Put performs a PUT request
func (c *HTTPClient) Put(ctx context.Context, endpoint string, request interface{}, response interface{}) error {
	url := c.baseURL + endpoint
	
	var body io.Reader
	if request != nil {
		jsonData, err := json.Marshal(request)
		if err != nil {
			return fmt.Errorf("failed to marshal request: %w", err)
		}
		body = bytes.NewBuffer(jsonData)
	}
	
	req, err := http.NewRequestWithContext(ctx, "PUT", url, body)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	
	if request != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	
	return c.doRequestWithRetry(req, response)
}

// Delete performs a DELETE request
func (c *HTTPClient) Delete(ctx context.Context, endpoint string, response interface{}) error {
	url := c.baseURL + endpoint
	
	req, err := http.NewRequestWithContext(ctx, "DELETE", url, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	
	return c.doRequestWithRetry(req, response)
}

// SetAuthToken sets the authentication token
func (c *HTTPClient) SetAuthToken(token string) {
	c.authToken = token
}

// Close closes the HTTP client
func (c *HTTPClient) Close() error {
	// HTTP client doesn't need explicit closing in Go
	return nil
}

// doRequestWithRetry performs a request with retry logic
func (c *HTTPClient) doRequestWithRetry(req *http.Request, response interface{}) error {
	var lastErr error
	
	for attempt := 0; attempt <= c.maxRetries; attempt++ {
		if attempt > 0 {
			delay := time.Duration(float64(c.retryDelay) * (1 << (attempt - 1)) * c.backoff)
			time.Sleep(delay)
		}
		
		err := c.doRequest(req, response)
		if err == nil {
			return nil
		}
		
		lastErr = err
		
		// Don't retry on client errors (4xx)
		if httpErr, ok := err.(*verinode.Error); ok && httpErr.StatusCode >= 400 && httpErr.StatusCode < 500 {
			break
		}
	}
	
	return lastErr
}

// doRequest performs a single HTTP request
func (c *HTTPClient) doRequest(req *http.Request, response interface{}) error {
	// Set headers
	req.Header.Set("User-Agent", "verinode-sdk-go/1.0.0")
	if c.authToken != "" {
		req.Header.Set("Authorization", "Bearer "+c.authToken)
	}
	
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return verinode.NewNetworkError(fmt.Sprintf("request failed: %v", err))
	}
	defer resp.Body.Close()
	
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return verinode.NewNetworkError(fmt.Sprintf("failed to read response body: %v", err))
	}
	
	// Try to parse as JSON
	var apiResponse map[string]interface{}
	if err := json.Unmarshal(body, &apiResponse); err != nil {
		return verinode.NewAPIError(fmt.Sprintf("invalid JSON response: %v", err), resp.StatusCode, string(body))
	}
	
	// Check for API error
	if success, ok := apiResponse["success"].(bool); ok && !success {
		errorMsg := "API error"
		if errMsg, ok := apiResponse["error"].(string); ok {
			errorMsg = errMsg
		}
		return verinode.NewAPIError(errorMsg, resp.StatusCode, apiResponse)
	}
	
	// Parse response if provided
	if response != nil {
		if err := json.Unmarshal(body, response); err != nil {
			return verinode.NewAPIError(fmt.Sprintf("failed to parse response: %v", err), resp.StatusCode, string(body))
		}
	}
	
	return nil
}

File: sdks/rust/src/utils/http.rs
//! HTTP client utilities for the Verinode SDK.

use crate::config::Config;
use crate::error::{Error, Result};
use async_trait::async_trait;
use reqwest::{Client, StatusCode};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::Duration;

/// Trait for HTTP client implementations
#[async_trait]
pub trait HttpClient: Send + Sync {
    /// Set the authentication token
    fn set_auth_token(&self, token: &str);
    
    /// Perform a GET request
    async fn get<T: for<'de> Deserialize<'de>>(&self, endpoint: &str) -> Result<T>;
    
    /// Perform a GET request with parameters
    async fn get_with_params<T: for<'de> Deserialize<'de>>(
        &self,
        endpoint: &str,
        params: &[(&String, serde_json::Value)],
    ) -> Result<T>;
    
    /// Perform a POST request
    async fn post<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        endpoint: &str,
        request: &Req,
    ) -> Result<Res>;
    
    /// Perform a PATCH request
    async fn patch<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        endpoint: &str,
        request: &Req,
    ) -> Result<Res>;
    
    /// Perform a PUT request
    async fn put<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        endpoint: &str,
        request: &Req,
    ) -> Result<Res>;
    
    /// Perform a DELETE request
    async fn delete<T: for<'de> Deserialize<'de>>(&self, endpoint: &str) -> Result<T>;
}

/// Reqwest-based HTTP client implementation
pub struct ReqwestHttpClient {
    client: Client,
    base_url: String,
    auth_token: std::sync::RwLock<Option<String>>,
    max_retries: u32,
    retry_delay: Duration,
    backoff_multiplier: f64,
}

impl ReqwestHttpClient {
    /// Create a new HTTP client
    pub fn new(config: Arc<Config>) -> Self {
        let timeout = config.timeout;
        let mut client_builder = Client::builder()
            .timeout(timeout)
            .user_agent("verinode-sdk-rust/1.0.0");
        
        // Configure TLS if needed
        client_builder = client_builder.use_rustls_tls();
        
        let client = client_builder.build().unwrap_or_else(|_| {
            Client::builder()
                .timeout(timeout)
                .user_agent("verinode-sdk-rust/1.0.0")
                .build()
                .expect("Failed to create HTTP client")
        });
        
        Self {
            client,
            base_url: config.api_endpoint.clone(),
            auth_token: std::sync::RwLock::new(None),
            max_retries: config.max_retries,
            retry_delay: config.retry_delay,
            backoff_multiplier: config.backoff_multiplier,
        }
    }
    
    /// Build the full URL for an endpoint
    fn build_url(&self, endpoint: &str) -> String {
        format!("{}{}", self.base_url, endpoint)
    }
    
    /// Add authentication header to request
    fn add_auth_header(&self, request: reqwest::RequestBuilder) -> reqwest::RequestBuilder {
        let auth_token = self.auth_token.read().unwrap();
        if let Some(token) = auth_token.as_ref() {
            request.header("Authorization", format!("Bearer {}", token))
        } else {
            request
        }
    }
    
    /// Perform a request with retry logic
    async fn request_with_retry<F, T>(&self, request_fn: F) -> Result<T>
    where
        F: Fn() -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<T>> + Send>>,
    {
        let mut last_error = None;
        
        for attempt in 0..=self.max_retries {
            match request_fn().await {
                Ok(result) => return Ok(result),
                Err(err) => {
                    last_error = Some(err.clone());
                    
                    // Don't retry on client errors (4xx)
                    if let Some(http_err) = err.downcast_ref::<Error>() {
                        if let Some(status) = http_err.status_code {
                            if status.is_client_error() {
                                break;
                            }
                        }
                    }
                    
                    // Don't retry on the last attempt
                    if attempt == self.max_retries {
                        break;
                    }
                    
                    // Calculate delay with exponential backoff
                    let delay = self.retry_delay * self.backoff_multiplier.powi(attempt as i32);
                    tokio::time::sleep(delay).await;
                }
            }
        }
        
        Err(last_error.unwrap_or_else(|| Error::network_error("Request failed after retries")))
    }
    
    /// Handle HTTP response
    async fn handle_response<T: for<'de> Deserialize<'de>>(
        &self,
        response: reqwest::Response,
    ) -> Result<T> {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        
        if status.is_success() {
            serde_json::from_str(&text).map_err(|e| {
                Error::api_error(
                    format!("Failed to parse response: {}", e),
                    status,
                )
            })
        } else {
            // Try to parse error response
            if let Ok(error_response) = serde_json::from_str::<serde_json::Value>(&text) {
                if let Some(error_msg) = error_response.get("error").and_then(|v| v.as_str()) {
                    return Err(Error::api_error(error_msg.to_string(), status));
                }
            }
            
            Err(Error::api_error(
                format!("HTTP {}: {}", status.as_u16(), text),
                status,
            ))
        }
    }
}

#[async_trait]
impl HttpClient for ReqwestHttpClient {
    fn set_auth_token(&self, token: &str) {
        let mut auth_token = self.auth_token.write().unwrap();
        *auth_token = Some(token.to_string());
    }
    
    async fn get<T: for<'de> Deserialize<'de>>(&self, endpoint: &str) -> Result<T> {
        self.request_with_retry(|| {
            Box::pin(async {
                let url = self.build_url(endpoint);
                let request = self.add_auth_header(self.client.get(&url));
                let response = request.send().await.map_err(Error::from)?;
                self.handle_response(response).await
            })
        }).await
    }
    
    async fn get_with_params<T: for<'de> Deserialize<'de>>(
        &self,
        endpoint: &str,
        params: &[(&String, serde_json::Value)],
    ) -> Result<T> {
        self.request_with_retry(|| {
            Box::pin(async {
                let url = self.build_url(endpoint);
                let mut request = self.add_auth_header(self.client.get(&url));
                
                // Add query parameters
                for (key, value) in params {
                    request = request.query(&[(key.as_str(), &value.to_string())]);
                }
                
                let response = request.send().await.map_err(Error::from)?;
                self.handle_response(response).await
            })
        }).await
    }
    
    async fn post<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        endpoint: &str,
        request: &Req,
    ) -> Result<Res> {
        self.request_with_retry(|| {
            Box::pin(async {
                let url = self.build_url(endpoint);
                let request = self.add_auth_header(self.client.post(&url).json(request));
                let response = request.send().await.map_err(Error::from)?;
                self.handle_response(response).await
            })
        }).await
    }
    
    async fn patch<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        endpoint: &str,
        request: &Req,
    ) -> Result<Res> {
        self.request_with_retry(|| {
            Box::pin(async {
                let url = self.build_url(endpoint);
                let request = self.add_auth_header(self.client.patch(&url).json(request));
                let response = request.send().await.map_err(Error::from)?;
                self.handle_response(response).await
            })
        }).await
    }
    
    async fn put<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        endpoint: &str,
        request: &Req,
    ) -> Result<Res> {
        self.request_with_retry(|| {
            Box::pin(async {
                let url = self.build_url(endpoint);
                let request = self.add_auth_header(self.client.put(&url).json(request));
                let response = request.send().await.map_err(Error::from)?;
                self.handle_response(response).await
            })
        }).await
    }
    
    async fn delete<T: for<'de> Deserialize<'de>>(&self, endpoint: &str) -> Result<T> {
        self.request_with_retry(|| {
            Box::pin(async {
                let url = self.build_url(endpoint);
                let request = self.add_auth_header(self.client.delete(&url));
                let response = request.send().await.map_err(Error::from)?;
                self.handle_response(response).await
            })
        }).await
    }
}

/// Mock HTTP client for testing
#[cfg(test)]
pub struct MockHttpClient {
    // Implementation would go here
}

#[cfg(test)]
impl MockHttpClient {
    pub fn new() -> Self {
        Self {}
    }
}

#[cfg(test)]
#[async_trait]
impl HttpClient for MockHttpClient {
    fn set_auth_token(&self, _token: &str) {
        // Mock implementation
    }
    
    async fn get<T: for<'de> Deserialize<'de>>(&self, _endpoint: &str) -> Result<T> {
        // Mock implementation
        Err(Error::network_error("Not implemented"))
    }
    
    async fn get_with_params<T: for<'de> Deserialize<'de>>(
        &self,
        _endpoint: &str,
        _params: &[(&String, serde_json::Value)],
    ) -> Result<T> {
        // Mock implementation
        Err(Error::network_error("Not implemented"))
    }
    
    async fn post<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        _endpoint: &str,
        _request: &Req,
    ) -> Result<Res> {
        // Mock implementation
        Err(Error::network_error("Not implemented"))
    }
    
    async fn patch<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        _endpoint: &str,
        _request: &Req,
    ) -> Result<Res> {
        // Mock implementation
        Err(Error::network_error("Not implemented"))
    }
    
    async fn put<Req: Serialize, Res: for<'de> Deserialize<'de>>(
        &self,
        _endpoint: &str,
        _request: &Req,
    ) -> Result<Res> {
        // Mock implementation
        Err(Error::network_error("Not implemented"))
    }
    
    async fn delete<T: for<'de> Deserialize<'de>>(&self, _endpoint: &str) -> Result<T> {
        // Mock implementation
        Err(Error::network_error("Not implemented"))
    }
}
